File: backend/app/lexproof/services/roles.py
"""Fixed organization roles. Values are stored as strings in Firestore."""

from __future__ import annotations

from enum import Enum


class OrgRole(str, Enum):
    ADMIN = "admin"
    CONTRACT_OWNER = "contract_owner"
    REVIEWER = "reviewer"
    APPROVER = "approver"
    AUDITOR = "auditor"


ALL_ROLES = tuple(role.value for role in OrgRole)
ADMIN_ROLES = (OrgRole.ADMIN.value,)
MEMBER_MANAGEMENT_ROLES = (OrgRole.ADMIN.value,)
WORKFLOW_DEFINITION_ROLES = (OrgRole.ADMIN.value,)


def normalize_roles(roles: list[str] | tuple[str, ...] | None) -> list[str]:
    unique: list[str] = []
    for role in roles or []:
        value = str(role).strip().lower()
        if value not in ALL_ROLES:
            raise ValueError(f"Unknown role: {role}")
        if value not in unique:
            unique.append(value)
    return unique


def has_any_role(held: list[str] | tuple[str, ...] | None, required: list[str] | tuple[str, ...] | None) -> bool:
    if not required:
        return True
    held_set = set(held or [])
    return bool(held_set.intersection(required))

File: frontend/lib/roles.ts
export type OrgRole = 'admin' | 'contract_owner' | 'reviewer' | 'approver' | 'auditor'

export const ALL_ROLES: OrgRole[] = ['admin', 'contract_owner', 'reviewer', 'approver', 'auditor']

export const ROLE_LABELS: Record<OrgRole, string> = {
  admin: 'Admin',
  contract_owner: 'Contract owner',
  reviewer: 'Reviewer',
  approver: 'Approver',
  auditor: 'Auditor',
}

export function isOrgRole(value: string): value is OrgRole {
  return ALL_ROLES.includes(value as OrgRole)
}

export function hasRole(roles: string[] | undefined, role: OrgRole): boolean {
  return (roles || []).includes(role)
}

export function isAdmin(roles: string[] | undefined): boolean {
  return hasRole(roles, 'admin')
}

export function canAccessEvaluation(roles: string[] | undefined): boolean {
  return hasRole(roles, 'admin') || hasRole(roles, 'reviewer') || hasRole(roles, 'auditor')
}
